src/org/jgroups/protocols/JDBC_PING2.java

```diff
@@ -153,6 +161,107 @@ public String dump(String cluster) throws Exception {
         return list.stream().map(pd -> String.format("%s", pd)).collect(Collectors.joining("\n"));
     }
 
+    @Override
+    protected void handleView(View new_view, View old_view, boolean coord_changed) {
+        // If we are the coordinator, it is good to learn about new entries that have been added before we delete them.
+        // If we are not the coordinator, it is good to learn the new entries added by the coordinator.
+        // This avoids a "JGRP000032: %s: no physical address for %s, dropping message" that leads to split clusters at concurrent startup.
+        learnExistingAddresses();
+
+        // This is an updated logic where we do not call removeAll but instead remove those obsolete entries.
+        // This avoids the short moment where the table is empty and a new node might not see any other node.
+        if (is_coord) {
+            if (remove_old_coords_on_view_change) {
+                Address old_coord = old_view != null ? old_view.getCreator() : null;
+                if (old_coord != null)
+                    remove(cluster_name, old_coord);
+            }
+            Address[] left = View.diff(old_view, new_view)[1];
+            if (coord_changed || update_store_on_view_change || left.length > 0) {
+                writeAll(left);
+                if (remove_all_data_on_view_change) {
+                    removeAllNotInCurrentView();
+                }
+                if (remove_all_data_on_view_change || remove_old_coords_on_view_change) {
+                    startInfoWriter();
+                }
+            }
+        } else if (coord_changed && !remove_all_data_on_view_change) {
+            // I'm no longer the coordinator, usually due to a merge.
+            // The new coordinator will update my status to non-coordinator, and remove me fully
+            // if 'remove_all_data_on_view_change' is enabled and I'm no longer part of the view.
+            // Maybe this branch even be removed completely, but for JDBC_PING 'remove_all_data_on_view_change' is always set to true.
+            writeLocalAddress();
+        }
+    }
+
+    protected void removeAllNotInCurrentView() {
+       try (var conn = getConnection()) {
+          List<PingData> list = readFromDB(getClusterName());
+          for (PingData data : list) {
+             Address addr = data.getAddress();
+             if (view != null && !view.containsMember(addr)) {
+                addDiscoveryResponseToCaches(addr, data.getLogicalName(), data.getPhysicalAddr());
+                delete(conn, cluster_name, addr);
+             }
+          }
+       } catch (Exception e) {
+          log.error(String.format("%s: failed reading from the DB", local_addr), e);
+       }
```

Comment:
if you want to avoid reading the volatile field inside the loop: 
```suggestion
        View local_view = view;
        if (local_view == null) {
            return;
        }
        String cluster_name = getClusterName();
        try (var conn = getConnection()) {
            List<PingData> list = readFromDB(getClusterName());
            for (PingData data : list) {
                Address addr = data.getAddress();
                if (local_view.containsMember(addr)) {
                    continue;
                }
                addDiscoveryResponseToCaches(addr, data.getLogicalName(), data.getPhysicalAddr());
                delete(conn, cluster_name, addr);
            }
        } catch (Exception e) {
            log.error(String.format("%s: failed reading from the DB", local_addr), e);
        }
```

Reply:
Thanks. I modified your suggestion to remove `continue` in favour of a false assertion in the if statement:

```java
                if (!local_view.containsMember(addr)) {
                    addDiscoveryResponseToCaches(addr, data.getLogicalName(), data.getPhysicalAddr());
                    delete(conn, cluster_name, addr);
                }
```